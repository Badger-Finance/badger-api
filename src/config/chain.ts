import { ethers } from 'ethers';
import { SettDefinition } from '../interface/Sett';
import { Token } from '../interface/Token';
import { Provider } from './constants';
import { ethSetts } from './setts/eth-setts';

export interface Chain {
  readonly name: string;
  readonly symbol: string;
  readonly chainId: string;
  readonly tokens: Token[];
  readonly setts: SettDefinition[];
  readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
  public name = 'Ethereum';
  public symbol = 'eth';
  public chainId = '0x01';
  public tokens = [];
  public setts = ethSetts;
  public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
  public name = 'BinanceSmartChain';
  public symbol = 'bsc';
  public chainId = '0x61';
  public tokens = [];
  public setts = [];
  public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 * TODO: Should not supported throw a bad request instead?
 */
export const resolveChainQuery = (symbol: string): Chain => {
  if (!symbol) return eth;
  const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
  if (!chain) return eth;
  return chain;
};
