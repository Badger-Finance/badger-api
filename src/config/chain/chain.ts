import { BadRequest } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { SettDefinition } from '../../interface/Sett';
import { Token } from '../../interface/Token';
import { Provider } from '../constants';
import { bscSetts } from './bsc';
import { ethSetts } from './eth';

export interface Chain {
	readonly name: string;
	readonly symbol: string;
	readonly chainId: string;
	readonly tokens: Token[];
	readonly setts: SettDefinition[];
	readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
	public name = 'Ethereum';
	public symbol = 'eth';
	public chainId = '0x01';
	public tokens = [];
	public setts = ethSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
	public name = 'BinanceSmartChain';
	public symbol = 'bsc';
	public chainId = '0x61';
	public tokens = [];
	public setts = bscSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 */
export const resolveChainQuery = (symbol: string): Chain => {
	if (!symbol) return eth;
	const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
	if (!chain) throw new BadRequest('Invalid chain');
	return chain;
};
