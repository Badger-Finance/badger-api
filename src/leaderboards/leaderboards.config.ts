import { BadgerType } from './enums/badger-type.enum';
import { BadgerTypeMap } from './types/badger-type-map';

export const BADGER_RANKS: BadgerTypeMap = {
  [BadgerType.Basic]: 1,
  [BadgerType.Neo]: 20,
  [BadgerType.Hero]: 200,
  [BadgerType.Hyper]: 600,
  [BadgerType.Frenzy]: 1400,
};

export function getBadgerType(score: number): BadgerType {
  if (score >= BADGER_RANKS[BadgerType.Frenzy]) {
    return BadgerType.Frenzy;
  }
  if (score >= BADGER_RANKS[BadgerType.Hyper]) {
    return BadgerType.Hyper;
  }
  if (score >= BADGER_RANKS[BadgerType.Hero]) {
    return BadgerType.Hero;
  }
  if (score >= BADGER_RANKS[BadgerType.Neo]) {
    return BadgerType.Neo;
  }
  return BadgerType.Basic;
}
