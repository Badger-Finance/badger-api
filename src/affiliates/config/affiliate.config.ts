import { BadRequest } from '@tsed/exceptions';
import { Chain } from '../../chains/config/chain.config';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';

type Affiliates = { [protocol: string]: Affiliate };

export abstract class Affiliate {
  private static affiliates: Affiliates = {};

  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  static register(protocol: Protocol, affiliate: Affiliate): void {
    this.affiliates[protocol] = affiliate;
  }

  static getAffiliate(affiliateData: SettAffiliateData): Affiliate {
    const affiliate = this.affiliates[affiliateData.protocol];
    if (!affiliate) {
      throw new BadRequest(`${affiliate} is not a affiliate`);
    }
    return affiliate;
  }

  abstract getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData>;
}
