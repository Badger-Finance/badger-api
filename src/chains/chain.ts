import { Arbitrum } from './config/arbitrum.config';
import { BinanceSmartChain } from './config/bsc.config';
import { Chain } from './config/chain.config';
import { Ethereum } from './config/eth.config';
import { Polygon } from './config/matic.config';

/**
 * Instantiate objects for registration.
 */
export const loadChains = (): Chain[] => {
  return [new Ethereum(), new BinanceSmartChain(), new Polygon(), new Arbitrum()];
};
