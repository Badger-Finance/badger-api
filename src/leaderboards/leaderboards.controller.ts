import { LeaderboardSummary, Network } from "@badger-dao/sdk";
import { Controller, Get, Inject, QueryParams } from "@tsed/common";
import { ContentType, Hidden } from "@tsed/schema";

import { Chain } from "../chains/config/chain.config";
import { LeaderBoardsService } from "./leaderboards.service";

@Controller("/leaderboards")
export class LeaderBoardsController {
  @Inject()
  leaderBoardsService!: LeaderBoardsService;

  @Get("")
  @Hidden()
  @ContentType("json")
  async getLeaderBoardSummary(@QueryParams("chain") chain?: Network): Promise<LeaderboardSummary> {
    return this.leaderBoardsService.fetchLeaderboardSummary(Chain.getChain(chain));
  }
}
