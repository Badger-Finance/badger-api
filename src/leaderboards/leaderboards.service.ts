import { BadgerTypeMap, LeaderboardSummary } from '@badger-dao/sdk';
import { Service } from '@tsed/common';

import { Chain } from '../chains/config/chain.config';
import { queryLeaderboardSummary } from './leaderboards.utils';

@Service()
export class LeaderBoardsService {
  async fetchLeaderboardSummary(chain: Chain): Promise<LeaderboardSummary> {
    const cachedSummary = await queryLeaderboardSummary(chain);
    const summary = Object.fromEntries(
      cachedSummary.rankSummaries.map((s) => [s.badgerType, s.amount])
    ) as BadgerTypeMap;
    return {
      summary,
      updatedAt: cachedSummary.updatedAt
    };
  }
}
