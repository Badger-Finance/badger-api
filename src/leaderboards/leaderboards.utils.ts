import { BadgerType } from '@badger-dao/sdk';
import { ethers } from 'ethers';
import { getDataMapper } from '../aws/dynamodb.utils';
import { Chain } from '../chains/config/chain.config';
import { getLeaderboardKey } from '../indexers/leaderboard-indexer';
import { CachedBoost } from './interface/cached-boost.interface';
import { CachedLeaderboardSummary } from './interface/cached-leaderboard-summary.interface';

export async function queryLeaderboardSummary(chain: Chain): Promise<CachedLeaderboardSummary> {
  const mapper = getDataMapper();
  for await (const entry of mapper.query(
    CachedLeaderboardSummary,
    {
      leaderboard: getLeaderboardKey(chain),
    },
    { limit: 1 },
  )) {
    return entry;
  }
  return {
    leaderboard: getLeaderboardKey(chain),
    rankSummaries: [
      {
        badgerType: BadgerType.Basic,
        amount: 0,
      },
      {
        badgerType: BadgerType.Neo,
        amount: 0,
      },
      {
        badgerType: BadgerType.Hero,
        amount: 0,
      },
      {
        badgerType: BadgerType.Hyper,
        amount: 0,
      },
      {
        badgerType: BadgerType.Frenzy,
        amount: 0,
      },
    ],
    updatedAt: Date.now(),
  };
}

export async function getLeaderBoardSize(chain: Chain): Promise<number> {
  const mapper = getDataMapper();
  for await (const entry of mapper.query(
    CachedBoost,
    {
      leaderboard: getLeaderboardKey(chain),
    },
    { limit: 1, scanIndexForward: false },
  )) {
    return entry.rank;
  }
  return 0;
}

export async function getUserLeaderBoardRank(chain: Chain, accountId: string): Promise<number> {
  const mapper = getDataMapper();
  for await (const entry of mapper.query(
    CachedBoost,
    { leaderboard: getLeaderboardKey(chain), address: ethers.utils.getAddress(accountId) },
    { limit: 1, indexName: 'IndexLeaderBoardRankOnAddress' },
  )) {
    return entry.rank;
  }
  const leaderboardSize = await getLeaderBoardSize(chain);
  return leaderboardSize + 1;
}
