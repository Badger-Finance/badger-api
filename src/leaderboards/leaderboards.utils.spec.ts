import { BadgerType } from '@badger-dao/sdk';

import { getLeaderboardKey } from '../aws/dynamodb.utils';
import { Chain } from '../chains/config/chain.config';
import { mockQuery, setupMockChain } from '../test/mocks.utils';
import { queryLeaderboardSummary } from './leaderboards.utils';

describe('leaderboards.utils', () => {
  describe('queryLeaderboardSummary', () => {
    let chain: Chain;

    beforeEach(() => {
      chain = setupMockChain();
    });

    describe('no saved leaderboard summary data', () => {
      it('returns a map of all badger ranks with zero entries', async () => {
        mockQuery([]);
        const result = await queryLeaderboardSummary(chain);
        // result date will always update due to nature of function
        result.updatedAt = 133742069;
        expect(result).toMatchSnapshot();
      });
    });

    describe('saved leaderboard summary data', () => {
      it('returns the appropriate TEST_CHAIN summary data', async () => {
        mockQuery([
          {
            leaderboard: getLeaderboardKey(chain.network),
            rankSummaries: [
              {
                badgerType: BadgerType.Basic,
                amount: 1000,
              },
              {
                badgerType: BadgerType.Neo,
                amount: 20,
              },
              {
                badgerType: BadgerType.Hero,
                amount: 35,
              },
              {
                badgerType: BadgerType.Hyper,
                amount: 25,
              },
              {
                badgerType: BadgerType.Frenzy,
                amount: 40,
              },
            ],
            updatedAt: 133742069,
          },
        ]);
        const result = await queryLeaderboardSummary(chain);
        expect(result).toMatchSnapshot();
      });
    });
  });
});
