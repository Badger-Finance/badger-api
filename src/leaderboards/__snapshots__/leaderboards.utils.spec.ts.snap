// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`leaderboards.utils queryLeaderboardSummary no saved leaderboard summary data returns a map of all badger ranks with zero entries 1`] = `
Object {
  "leaderboard": "ethereum_boost",
  "rankSummaries": Array [
    Object {
      "amount": 0,
      "badgerType": "basic",
    },
    Object {
      "amount": 0,
      "badgerType": "neo",
    },
    Object {
      "amount": 0,
      "badgerType": "hero",
    },
    Object {
      "amount": 0,
      "badgerType": "hyper",
    },
    Object {
      "amount": 0,
      "badgerType": "frenzy",
    },
  ],
  "updatedAt": 133742069,
}
`;

exports[`leaderboards.utils queryLeaderboardSummary saved leaderboard summary data returns the appropriate chain summary data 1`] = `
Object {
  "leaderboard": "ethereum_boost",
  "rankSummaries": Array [
    Object {
      "amount": 1000,
      "badgerType": "basic",
    },
    Object {
      "amount": 20,
      "badgerType": "neo",
    },
    Object {
      "amount": 35,
      "badgerType": "hero",
    },
    Object {
      "amount": 25,
      "badgerType": "hyper",
    },
    Object {
      "amount": 40,
      "badgerType": "frenzy",
    },
  ],
  "updatedAt": 133742069,
}
`;
